romi-rover/grbl/grbl.c

```diff
@@ -138,6 +155,73 @@ static int reset_cnc()
         return 0;
 }
 
+static int grbl_idle()
+{
+        int ret = -1;
+        const char *r;
+
+        r_debug("CNC: put '?'");
+        serial_lock(_serial);
+        serial_put(_serial, '?');
+        r = serial_readline(_serial, _reply);
+        serial_unlock(_serial);
+
+        if (r) r_debug("CNC status: '%s'", r);
+        else r_debug("CNC status: NULL!");
+
+        if (r == NULL)
+                ret = -1;
+        else if (r[0] != '<')
+                ret = -1;
+        else if (strlen(r) < 5)
+                ret = -1;
+        else if (strncmp(r+1, "Run", 3) == 0)
+                ret = 0;
+        else if (strncmp(r+1, "Idle", 4) == 0)
+                ret = 1;
+        else ret = 0;
+
+        return ret;
+}
+
+static void wait_cnc()
+{
+        double start_time = clock_time();
+
+        serial_lock(_serial);
+        serial_flush(_serial);
+        serial_unlock(_serial);
+
+        while (grbl_idle() != 1 && clock_time() - start_time < 600.0) {
+                r_debug("Waiting for CNC to finish");
+                clock_sleep(1);
+        }
+}
+
+static int moveto(int hasx, int hasy, int hasz,
```

Comment:
I'm not sure why this function is always inside the mutex then the send_serial_command_unlocked exists. Perhaps we could talk through the mutex strategy in this file sometime. It smells wrong to have a locked and unlocked function!